packages/backend-common/src/database/connection.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import knex from 'knex';
+import { Config } from '@backstage/config';
+import { buildPgConnectionConfig } from './postgres';
+import { buildSqlite3ConnectionConfig } from './sqlite3';
+
+/**
+ * Creates a knex database connection
+ *
+ * @param config The database config
+ * @param dbName The optional database to connect to
+ */
+export function createDatabase(config: Config, dbName?: string) {
+  const knexConfig = getDatabaseConfig(config, dbName);
+  const database = knex(knexConfig);
+
+  database.client.pool.on('createSuccess', (_eventId: any, resource: any) => {
+    resource.run('PRAGMA foreign_keys = ON', () => {});
+  });
+
+  return database;
+}
+
+/**
+ * Gets a knex database config
+ *
+ * @param config The database config
+ * @param dbName The optional database to connect to
+ */
+export function getDatabaseConfig(
+  config: Config,
+  dbName?: string,
+): knex.Config {
+  const client = config.getString('client');
+
+  return {
+    client,
+    connection: buildDatabaseConnectionConfig(client, config, dbName),
+    useNullAsDefault: true,
+  };
+}
+
+function buildDatabaseConnectionConfig(
+  client: string,
+  config: Config,
+  dbName?: string,
+): knex.PgConnectionConfig | knex.Sqlite3ConnectionConfig {
+  if (client === 'pg') {
+    return buildPgConnectionConfig(config, dbName);
+  }
+
+  return buildSqlite3ConnectionConfig(config);
```

Comment:
no else throw for unknown clients?

also, is there a use case for just leaving this open, reading the yaml "raw" for unknown clients at least? otherwise we have to update this package to explicitly support each database that people want to hook up.

Reply:
I agree. I think adding a case for known handlers makes sense and then maybe fallback to a straight pass through.

Reply:
Possibly with the exception of `PRAGMA foreign_keys = ON` - so it may warrant some special treatment (at some level) anyway.